pkg/apimachinery/registered/registered.go

```diff
@@ -42,75 +47,99 @@ var (
 	// subset of registeredVersions. Please call EnableVersions() to add
 	// enabled versions.
-	enabledVersions = map[unversioned.GroupVersion]struct{}{}
+	enabledVersions map[unversioned.GroupVersion]struct{}
 
 	// map of group meta for all groups.
-	groupMetaMap = map[string]*apimachinery.GroupMeta{}
+	groupMetaMap map[string]*apimachinery.GroupMeta
 
 	// envRequestedVersions represents the versions requested via the
 	// KUBE_API_VERSIONS environment variable. The install package of each group
 	// checks this list before add their versions to the latest package and
 	// Scheme.  This list is small and order matters, so represent as a slice
-	envRequestedVersions = []unversioned.GroupVersion{}
-)
-
-func init() {
-	loadKubeAPIVersions()
+	envRequestedVersions []unversioned.GroupVersion
 }
 
-func loadKubeAPIVersions() {
-	// Env var KUBE_API_VERSIONS is a comma separated list of API versions that
-	// should be registered in the scheme.
-	kubeAPIVersions := os.Getenv("KUBE_API_VERSIONS")
+// NewAPIRegistrationManager constructs a new manager. The argument ought to be
+// the value of the KUBE_API_VERSIONS env var, or a value of this which you
+// wish to test.
+func NewAPIRegistrationManager(kubeAPIVersions string) (*APIRegistrationManager, error) {
+	m := &APIRegistrationManager{
+		registeredVersions:      map[unversioned.GroupVersion]struct{}{},
+		thirdPartyGroupVersions: []unversioned.GroupVersion{},
+		enabledVersions:         map[unversioned.GroupVersion]struct{}{},
+		groupMetaMap:            map[string]*apimachinery.GroupMeta{},
+		envRequestedVersions:    []unversioned.GroupVersion{},
+	}
+
 	if len(kubeAPIVersions) != 0 {
 		for _, version := range strings.Split(kubeAPIVersions, ",") {
 			gv, err := unversioned.ParseGroupVersion(version)
 			if err != nil {
-				glog.Fatalf("invalid api version: %s in KUBE_API_VERSIONS: %s.",
-					version, os.Getenv("KUBE_API_VERSIONS"))
+				return nil, fmt.Errorf("invalid api version: %s in KUBE_API_VERSIONS: %s.",
+					version, kubeAPIVersions)
 			}
-			envRequestedVersions = append(envRequestedVersions, gv)
+			m.envRequestedVersions = append(m.envRequestedVersions, gv)
 		}
 	}
+	return m, nil
 }
 
-// Resets everything to clean room for the start of a test
-func clearForTesting() {
-	registeredVersions = map[unversioned.GroupVersion]struct{}{}
-	thirdPartyGroupVersions = []unversioned.GroupVersion{}
-	enabledVersions = map[unversioned.GroupVersion]struct{}{}
-	groupMetaMap = map[string]*apimachinery.GroupMeta{}
-	envRequestedVersions = []unversioned.GroupVersion{}
-	loadKubeAPIVersions()
+func NewOrDie(kubeAPIVersions string) *APIRegistrationManager {
+	m, err := NewAPIRegistrationManager(kubeAPIVersions)
+	if err != nil {
+		glog.Fatalf("Could not construct version manager: %v (KUBE_API_VERSIONS=%q)", err, kubeAPIVersions)
+	}
+	return m
 }
 
+// People are calling global functions. Let them continue to do that (for now).
```

Comment:
Not sure if there's an issue or not. I don't think it's easy work, unfortunately. :(

We *do* need a global registry for *available* apis, but this code should *not* be a global registry at all. Once we have the former, people will be able to make the latter at will. My https://github.com/kubernetes/kubernetes/pull/30565 is the next step.